strawberryfields/backends/states.py

```diff
@@ -1250,6 +1252,18 @@ def number_expectation(self, modes):
             "The number_expectation method only supports one or two modes for Gaussian states."
         )
 
+    def parity_expectation(self, modes):
+        """Calculates the expectation value of a product of parity operators acting on given modes"""
```

Comment:
Now that the ParityOperator will be supported by the `BaseGaussianState` class too, a more general `parity_expectation` method can be defined for the parent [`class BaseState(abc.ABC)`](https://github.com/XanaduAI/strawberryfields/blob/a9d12b986689377893f47a07a6fa1763cc1cd821/strawberryfields/backends/states.py#L37) (linking the locations pointing to the source code throughout this comment).

This way a shared docstring can be used here.

**Changes to be done**

For this, the following can be done:
1. Creating a `parity_expectation` method for the `BaseState` class for example [in between the `def number_expectation(self, modes):` and `def p_quad_values(self, mode, xvec, pvec):` definitions](https://github.com/XanaduAI/strawberryfields/blob/a9d12b986689377893f47a07a6fa1763cc1cd821/strawberryfields/backends/states.py#L368):
```python
    def parity_expectation(self, modes):
        """Calculates the expectation value of a product of parity operators acting on given modes"""
        raise NotImplementedError

```
2. Removing the [docstring of the `parity_expectation` method in the `BaseFockState` class](https://github.com/XanaduAI/strawberryfields/blob/a9d12b986689377893f47a07a6fa1763cc1cd821/strawberryfields/backends/states.py#L924) .
3. Removing the docstring of the `parity_expectation` method here too.

**More details on why this is useful**

Each state class inherits from the `BaseState` class (see for example the [`BaseFockState`](https://github.com/XanaduAI/strawberryfields/blob/a9d12b986689377893f47a07a6fa1763cc1cd821/strawberryfields/backends/states.py#L412)  and [`BaseGaussianState`](https://github.com/XanaduAI/strawberryfields/blob/a9d12b986689377893f47a07a6fa1763cc1cd821/strawberryfields/backends/states.py#L930) class which have `BaseState` in their definitions).

Class inheritance can be used not just for inheriting methods, but for inheriting docstrings too. Eventually, any class that inherits from `BaseState` will have a `parity_expectation` method defined (with the definition provided in step 1) this would simply be raising a `NotImplementedError`) and have the same shared docstring for the `parity_expectation` method.

So essentially here the the `parity_expectation` method would be uniquely defined for the `BaseGaussianState` class, but the docstring would be inherited from the `BaseState` class.

Reply:
Thanks for the detailed comments @antalszava! Would you prefer that these changes are made in a separate PR?

Reply:
Sure thing! It is up to you, I think it's not too crucial to be done here. A follow-up PR could include making both this change and the change with `number_expectation`.

Reply:
Sounds good, I'll make the changes and open a new PR tomorrow 🙂